### PR TITLE
Fix incorrect removal of VRF device

### PR DESCRIPTION
**Pull request by pperiyasamy**
This PR fixes repair method of VRF manager which is buggy and it incorrectly removes valid VRF device associated with a user defined network while running reconcile loop.

### CODE REVIEW

**Review by kyrtapz — DISMISSED** [review]
The change itself looks good to me but it feels like we are missing either a unit test or an E2E to make sure that the repair behaves as expected.

**Comment by pperiyasamy** [comment]
> but it feels like we are missing either a unit test

yes @kyrtapz , updated the unit test to cover this.

File: go-controller/pkg/node/vrfmanager/vrf_manager.go [hunk 1]
```diff
@@ -298,8 +298,8 @@ func (vrfm *Controller) repair(validVRFs sets.Set[string]) error {
 			if err != nil {
 				klog.Errorf("VRF Manager: error deleting stale VRF device %s, err: %v", name, err)
 			}
+			delete(vrfm.vrfs, link.Attrs().Index)
```
Comment (tssurya) [inline go-controller/pkg/node/vrfmanager/vrf_manager.go:301]: nit: yeah the other option was to continue if it was validVRFs.Has(name); neater for reading loop

File: go-controller/pkg/node/vrfmanager/vrf_manager_test.go [hunk 2]
```diff
@@ -116,6 +116,9 @@ var _ = ginkgo.Describe("VRF manager", func() {
 			enslaveLinkMock1.On("Type").Return("dummy")
 			err = c.reconcile()
 			gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+			// Invoke reconcile again to ensure both vrf links in sync.
+			err = c.reconcile()
+			gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
```
Comment (tssurya) [inline go-controller/pkg/node/vrfmanager/vrf_manager_test.go:121]: hmm I see you are testing error didn't happen... but shouldn't we check the vrf was not removed? I don't see the ensure vrf links in sync is happening...? what am I missing..

Comment (tssurya) [reply go-controller/pkg/node/vrfmanager/vrf_manager_test.go:121]: spoke with Peri, he will add another test to check reconcile+ vrf deletion's better

Comment (pperiyasamy) [reply go-controller/pkg/node/vrfmanager/vrf_manager_test.go:129]: yes @tssurya , added a test now. PTAL.

**Review by kyrtapz — APPROVED** [review]

File: go-controller/pkg/node/vrfmanager/vrf_manager_test.go [hunk 3]
```diff
@@ -130,3 +141,98 @@ var _ = ginkgo.Describe("VRF manager", func() {
...
+			wg3.Wait()
+
+			// Invoke reconcile method explicitly few times to ensure it's always working fine.
+			err = c.reconcile()
+			gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+			err = checkforVrfLinkExistence()
+			gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+			err = c.reconcile()
+			gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+			err = checkforVrfLinkExistence()
```
Comment (tssurya) [inline go-controller/pkg/node/vrfmanager/vrf_manager_test.go:231]: awesome!!

**Review by tssurya — APPROVED** [review]

**Pull request merged**